cfme/infrastructure/virtual_machines.py

```diff
@@ -104,6 +122,204 @@ def is_displayed(self):
+        vms = appl_db['vms']
+        hws = appl_db['hardwares']
+        hw_data = appl_db.session.query(ems, vms, hws).filter(
+            ems.name == self.vm.provider.name).filter(
+            vms.ems_id == ems.id).filter(
+            vms.name == self.vm.name).filter(
+            hws.vm_or_template_id == vms.id
+        ).first().hardwares
+        self.hw = VMHardware(
+            hw_data.cpu_cores_per_socket, hw_data.cpu_sockets, hw_data.memory_mb, 'MB')
+        hw_id = hw_data.id
+
+        # Disks
+        disks = appl_db['disks']
+        disks_data = appl_db.session.query(disks).filter(
+            disks.hardware_id == hw_id).filter(
+            disks.device_type == 'disk'
+        ).all()
+        for disk_data in disks_data:
+            # In DB stored in bytes, but UI default is GB
+            size_gb = disk_data.size / (1024 ** 3)
+            self.disks.append(
+                VMDisk(
+                    filename=disk_data.filename,
+                    size=size_gb,
+                    size_unit='GB',
+                    type=disk_data.disk_type,
+                    mode=disk_data.mode
+                ))
+
+    def copy(self):
+        """Returns a copy of this configuration
+        """
+        config = VMConfiguration.__new__(VMConfiguration)
+        config.hw = copy(self.hw)
+        # We can just make shallow copy here because disks can be only added or deleted, not edited
+        config.disks = self.disks[:]
+        config.vm = self.vm
+        return config
+
+    def add_disk(self, size, size_unit='GB', type='thin', mode='persistent'):
+        """Adds a disk to the VM
+
+        Args:
+            size: Size of the disk
+            size_unit: Unit of size ('MB' or 'GB')
+            type: Type of the disk ('thin' or 'thick')
+            mode: Mode of the disk ('persistent', 'nonpersistent' or 'independent_nonpersistent')
+
+        Note:
+            This method is designed to correspond with the DB, not with the UI.
+            In the UI, dependency is represented by a separate Yes / No option which is _incorrect_
+            design that we don't follow. Correctly, mode should be a selectbox of 3 items:
+            Persistent, Nonpersistent and Independent Nonpersistent.
+            Independent Persistent is an invalid combination that UI currently (5.8) allows.
+        """
+        # New disk doesn't have a filename, until actually added
+        disk = VMDisk(
+            filename=None, size=size, size_unit=size_unit, type=type, mode=mode)
+        self.disks.append(disk)
+
+    def delete_disk(self, filename=None, index=None):
+        """Removes a disk of given filename or index"""
+        if filename:
+            disk = [disk for disk in self.disks if disk.filename == filename][0]
+            self.disks.remove(disk)
+        elif index:
+            del self.disks[index]
+        else:
+            raise ValueError("Either filename or index must be specified")
```

Comment:
This should be a TypeError - that is what happens when you call a function with wrong number of arguments.